tests/test_single_topology_v3.py

```diff
@@ -0,0 +1,180 @@
+        assert jks == [(2, None)]
+
+
+def testing_find_dummy_groups_and_multiple_anchors():
+    """
+    Test that we can find anchors and dummy groups with multiple anchors, we expect to find only a single
+    root anchor and neighbor core atom pair.
+    """
+    mol_a = Chem.MolFromSmiles("OCC")
+    mol_b = Chem.MolFromSmiles("O1CC1")
+    core = np.array([[1, 1], [2, 2]])
+
+    with pytest.warns(MultipleAnchorWarning):
+        dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core[:, 0], core[:, 1])
+        assert dgs == [{0}]
+        assert jks == [(1, 2)] or jks == [(2, 1)]
+
+    mol_a = Chem.MolFromSmiles("C(C)(C)C")
+    mol_b = Chem.MolFromSmiles("O1CCCC1")
+
+    core_a = [0, 1, 2, 3]
+    core_b = [2, 1, 4, 3]
+
+    with pytest.warns(MultipleAnchorWarning):
+        dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core_a, core_b)
+        assert dgs == [{0}]
+        assert jks == [(1, 2)]
+
+
+def compute_distance_matrix(x0):
+    N = x0.shape[0]
+    D = x0.shape[1]
+    xi = x0.reshape((1, N, D))
+    xj = x0.reshape((N, 1, D))
+    rij = xi - xj
+    dij = jnp.linalg.norm(rij, axis=-1)
+    return dij
+
+
+def get_max_distance(x0):
+    dij = compute_distance_matrix(x0)
+    return jnp.amax(dij)
+
+
+def test_hif2a_end_state_stability():
+    """
+    Pick some random pairs from the hif2a set and ensure that they're numerically stable at the
+    end-states under some a standard distance based atom_mapping protocol.
+    """
+    suppl = Chem.SDMolSupplier("timemachine/testsystems/data/ligands_40.sdf", removeHs=False)
+    mols = [m for m in suppl]
+    pairs = []
+    for mol_a in mols:
+        for mol_b in mols:
+            pairs.append((mol_a, mol_b))
+
+    np.random.seed(2023)
+    np.random.shuffle(pairs)
+    ff = Forcefield.load_from_file("smirnoff_1_1_0_sc.py")
+
+    batch_distance_check = jax.vmap(get_max_distance)
+
+    # this has been tested for up to 50 random pairs
+    for mol_a, mol_b in pairs[:5]:
+
+        print("Checking", mol_a.GetProp("_Name"), "->", mol_b.GetProp("_Name"))
+        mcs_threshold = 0.75  # distance threshold, in nanometers
+        res = atom_mapping.mcs_map(mol_a, mol_b, mcs_threshold)
+        query = Chem.MolFromSmarts(res.smartsString)
+        core = atom_mapping.get_core_by_mcs(mol_a, mol_b, query, mcs_threshold)
+        st = SingleTopologyV3(mol_a, mol_b, core, ff)
+        x0 = st.combine_confs(get_romol_conf(mol_a), get_romol_conf(mol_b))
+        systems = [
+            st.setup_end_state_src(),
+            st.setup_end_state_dst(),
+        ]
+
+        for system in systems:
+            U_fn = system.get_U_fn()
+            batch_U_fn = jax.vmap(U_fn)
+            frames = simulate_system(system.get_U_fn(), x0, num_samples=1000)
+            nrgs = batch_U_fn(frames)
+
+            assert np.all(np.isfinite(nrgs))
+            assert np.all(np.isfinite(frames))
+            distance_cutoff = 2.5  # in nanometers
```

Comment:
(Possible refinement for later: could tighten this `distance_cutoff` to `max([length(path) for path in shortest_paths(mol_graph)])` or similar)

Reply:
agreed - but saving this for later.